### ISSUE TITLE
Remote methods mapping in JSON

### ISSUE THREAD

**Issue by nikita-leonov** [strongloop/loopback#209]
Would be nice to have an option to define remote methods mapping to an actual classes in JSON. So we will have complete REST interface defined in one place with a logic and complex cases in JS. I do not see why some of such constructions could not be a part of JSON mapping:
ApplicationModel.authenticate,
    {
      accepts: [
        {arg: 'appId', type: 'string', required: true},
        {arg: 'key', type: 'string', required: true}          
      ],
      returns: {arg: 'accessToken', type: 'object', root: true},
      http: {verb: 'post'}
    }

**Comment by mrfelton** [comment]
fwiw I'm doing something like this in my own setup by using the unimplemented `methods` key in the model json schema to define the remote methods in the same way as you would normally in the `remoteMethod()` method, and then using that to setup the remote methods on the fly.

eg

common/models/my-model.json
```
  ...
  "relations": [],
  "acls": [],
  "methods": {
    "syncToRemote": {
      "doc": "Syncronise a person to a supported provider account.",
      "accepts": [{
        "arg": "id",
        "type": "number",
        "required": true,
        "description": "PersistedModel id"
      }, {
        "arg": "provider",
        "type": "string",
        "required": true,
        "http": {
          "source": "form"
        },
        "description": "The external account provider to update."
      }],
      "returns": {
        "arg": "person",
        "type": "object",
        "root": true,
        "description": "The updated person object."
      },
      "http": {
        "path": "/:id/syncToRemote",
        "verb": "post"
      }
    }
  }
```

common/models/my-model.js:
```
  // Set up remote methods from model config schema json.
  // TODO: Remove / Refactor this once loopback supports this natively.
  MyModel.on('attached', function () {
    var app = MyModel.app;
    _.each(app.models.MyModel.settings.methods, function (value, key) {
      MyModel.remoteMethod(key, value);
    });
  });
```

Seems to be working ok so far. My only gripe with it is that the remoteMethod call expects documentation for the properties to be in a 'description' key, which is inconsistent with all the rest of the documentation in common/models/my-model.json which uses the `doc` key instead.

Would be great to have some support like this in loopback itself.

**Comment by bajtos** [comment]
@mrfelton can you submit a pull request with your implementation? Remote methods should be defined by [registry.createModel](https://github.com/strongloop/loopback/blob/d5b7447750c4ca881744f5f9ce8be559a993d2ab/lib/registry.js#L87).

Few things to change in the second snippet:
 
1) I believe you don't have wait for the model to be attached. In LoopBack 2.x, the remoting metadata is stored in a name-based lookup table, it's ok to describe a remote method before it's mixed into the object.

2) The API of `Model.remoteMethod()` is suboptimal and we would like to fix it in 3.0, see https://github.com/strongloop/loopback/issues/597. I would like the JSON-based registration to use the new way, meaning that the method name "ping" should be mapped to a static method (`isStatic: true`) and the method name "prototype.pong" should be mapped to a prototype method (`isStatic: false`). This is something you have to handle in your code until `Model.remoteMethod()` is fixed.

> My only gripe with it is that the remoteMethod call expects documentation for the properties to be in a 'description' key, which is inconsistent with all the rest of the documentation in common/models/my-model.json which uses the 'doc' key instead.

The "doc" keyword is deprecated, you should use "description" in your model definitions too. See [Model definitions JSON file](http://docs.strongloop.com/display/LB/Model+definition+JSON+file#ModeldefinitionJSONfile-Top-levelproperties) docs.

**Comment by mrfelton** [comment]
Thanks @bajtos .

> 1) I believe you don't have wait for the model to be attached. In LoopBack 2.x, the remoting metadata is stored in a name-based lookup table, it's ok to describe a remote method before it's mixed into the object.

The reason I wait until the model is attached is because I needed access to the app object so that I can inspect the structure and pull out the method details from there. I wasn't able to get a handle on a complete enough app object anywhere else. I'm guessing that if I move it to`registry.createModel.` I'll have direct access to it from there.

> 2) The API of Model.remoteMethod() is suboptimal and we would like to fix it in 3.0, see #597. I would like the JSON-based registration to use the new way, meaning that the method name "ping" should be mapped to a static method (isStatic: true) and the method name "prototype.pong" should be mapped to a prototype method (isStatic: false).

I'm not sure what you mean by ping and pong. Any docs on that you can point me towards?

> The "doc" keyword is deprecated, you should use "description" in your model definitions too. See Model definitions JSON file docs.

Awesome!

**Comment by bajtos** [comment]
> > 2) The API of Model.remoteMethod() is suboptimal and we would like to fix it in 3.0, see #597. I would like the JSON-based registration to use the new way, meaning that the method name "ping" should be mapped to a static method (isStatic: true) and the method name "prototype.pong" should be mapped to a prototype method (isStatic: false).

> I'm not sure what you mean by ping and pong. Any docs on that you can point me towards?

They are just example method names. "ping" is a static method, "pong" is a prototype (instance) method.

The code sample in your comment above requires developers to register them in this way:

```js
remoteMethods: {
  "ping": { isStatic: true, /*...*/ },
  "pong": { isStatic: false, /*...*/ }
}
```

What I would like to see instead:

```js
remoteMethods: {
  "ping": { /*...*/ },
  "prototype.pong": { /*...*/ }
}
```

Does it make sense now?

**Comment by mrfelton** [comment]
gotcha. I read through #597 that you lined to also so it makes more since to me now. I actually hadn't come across the isStatic property before.
